src/core/BaseUtilities.ts

```diff
@@ -568,17 +577,74 @@ export class BaseUtilities {
         }
     }
 
-    readonly #ensuredGuilds = new WeakSet<Guild>();
-    public async ensureMemberCache(guild: Guild): Promise<void> {
-        if (this.#ensuredGuilds.has(guild))
-            return;
+    readonly #guildsWithResolvedMembers = new WeakMap<Guild, Promise<void>>();
+    public ensureMemberCache(guild: Guild): Promise<void> {
+        let resolve = this.#guildsWithResolvedMembers.get(guild);
+        if (resolve === undefined) {
+            this.#guildsWithResolvedMembers.set(guild, resolve = this.#ensureMemberCache(guild).catch(err => {
+                if (err instanceof DiscordRESTError)
+                    this.#guildsWithResolvedMembers.delete(guild);
+                throw err;
+            }));
+        }
+        return resolve;
+    }
 
-        this.#ensuredGuilds.add(guild);
+    async #ensureMemberCache(guild: Guild): Promise<void> {
         const initialSize = guild.members.size;
         await guild.fetchAllMembers();
         this.logger.info('Cached', guild.members.size - initialSize, 'members in guild', guild.id, '. Member cache now has', guild.members.size, 'entries');
     }
 
+    public async * streamAllBans(guild: Guild): AsyncGenerator<GuildBan, void, undefined> {
+        let batch = [];
+        let after;
+        const bans = this.getGuildBans(guild);
+        do {
+            batch = await guild.getBans({ after });
+            for (const ban of batch)
+                bans.add(ban.user.id);
+            yield* batch;
+            after = batch.pop()?.user.id;
+        } while (after !== undefined);
+    }
+
+    public async requestAllBans(guild: Guild): Promise<GuildBan[]> {
+        const result = [];
+        for await (const ban of this.streamAllBans(guild))
+            result.push(ban);
+        return result;
+    }
+
+    readonly #guildsWithResolvedBans = new Map<string, Promise<void>>();
+    public ensureGuildBans(guild: Guild): Promise<void> {
+        const hasPerms = guild.members.get(this.user.id)?.permissions.has('banMembers') ?? false;
+        if (!hasPerms)
+            this.#guildsWithResolvedBans.delete(guild.id);
+        let resolve = this.#guildsWithResolvedBans.get(guild.id);
+        if (resolve === undefined) {
+            this.#guildsWithResolvedBans.set(guild.id, resolve = this.#ensureGuildBans(guild).catch(err => {
+                if (err instanceof DiscordRESTError)
+                    this.#guildsWithResolvedBans.delete(guild.id);
+                throw err;
+            }));
+        }
+        return resolve;
+    }
+
+    async #ensureGuildBans(guild: Guild): Promise<void> {
+        for await (const _ of this.streamAllBans(guild))
+            ;
+    }
```

Comment:
This makes sense as you simply want to cache the guild's bans... But it feels so weird 😆 Also, couldn't the semicolon be placed on the previous line ?

Reply:
fair point haha, Ill add a comment to make it clear why we exhaust the bans without using them